_posts/2025-02-20-agentic-ai-with-quarkus-p2.adoc

```diff
@@ -0,0 +1,312 @@
+
+== Agents
+
+Agents differ from the workflow patterns because the control flow, instead of being implemented programmatically, is entirely delegated to LLMs, which typically invoke tools to complete determinate subtasks. These tools can be seen as functions, opportunistically invoked by the LLM to achieve its goal and implemented in various technologies, like simple imperative methods, external web services, or even other LLMs.
+
+=== The weather forecast agent
+
+This https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/aiastool[first example] of agentic AI implements the following https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/aiastool/WeatherForecastAgent.java[weather forecast agent]
+
+[source, java]
+----
+@RegisterAiService(modelName = "tool-use")
+public interface WeatherForecastAgent {
+
+    @SystemMessage("""
+        You are a meteorologist, and you need to answer questions asked by the user about weather using at most 3 lines.
+
+        The weather information is a JSON object and has the following fields:
+
+        maxTemperature is the maximum temperature of the day in Celsius degrees
+        minTemperature is the minimum temperature of the day in Celsius degrees
+        precipitation is the amount of water in mm
+        windSpeed is the speed of wind in kilometers per hour
+        weather is the overall weather.
+    """)
+    @ToolBox({CityExtractorAgent.class, WeatherForecastService.class, GeoCodingService.class})
+    String chat(String query);
+}
+----
+
+The `@ToolBox` annotation allows to declaratively provide to this intelligent agent the three different tools that it will need to achieve its goal:
+
+1 . https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/aiastool/CityExtractorAgent.java[another LLM] specialized in extracting the name of a city from the user's prompt (thus also demonstrating how easily an LLM can be configured to become a tool for another LLM).
+
+[source, java]
+----
+@ApplicationScoped
+@RegisterAiService(chatMemoryProviderSupplier = RegisterAiService.NoChatMemoryProviderSupplier.class)
+public interface CityExtractorAgent {
+
+	@UserMessage("""
+    	You are given one question and you have to extract city name from it
+    	Only reply the city name if it exists or reply 'unknown_city' if there is no city name in question
+
+    	Here is the question: {question}
+    	""")
+	@Tool("Extracts the city from a question")
+	String extractCity(String question);
+}
+----
+
+2 .  a https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/aiastool/geo/GeoCodingService.java[web service] returning the geographic coordinates of a given city.
```

Comment:
Can you add a comment about how the use of a web service as a tool is significant? I can call the code as a web service, but when it's used as a Tool is it just treated the same way as any other Tool (where the agent framework calls the target method of the Tool directly)?